src/naip_cnn/data.py

```diff
@@ -1,160 +1,204 @@
 from __future__ import annotations
 
+from abc import ABC, abstractmethod
 from pathlib import Path
 
+import h5py
 import tensorflow as tf
 import tensorflow_io as tfio
 
 from naip_cnn.config import BANDS
 
 
-class Dataset:
-    """A Dataset of NAIP images and associated forest attribute labels, stored in H5."""
+@tf.autograph.experimental.do_not_convert
+def _augment(img: tf.Tensor, label: tf.Tensor) -> tuple[tf.Tensor, tf.Tensor]:
+    """Augment an image Tensor.
+
+    Parameters
+    ----------
+    img : tf.Tensor
+        The image Tensor to augment.
+    label : tf.Tensor
+        The label Tensor to augment.
+
+    Returns
+    -------
+    tf.Tensor
+        The augmented image Tensor.
+    """
+    flip_lr = tf.random.uniform([], 0, 1.0, dtype=tf.float32)
+    flip_ud = tf.random.uniform([], 0, 1.0, dtype=tf.float32)
+
+    # We need a channel dimension to use the tf.image functions
+    label = tf.expand_dims(label, axis=-1)
+
+    img = tf.cond(flip_lr > 0.5, lambda: tf.image.flip_left_right(img), lambda: img)
+    img = tf.cond(flip_ud > 0.5, lambda: tf.image.flip_up_down(img), lambda: img)
+    label = tf.cond(
+        flip_lr > 0.5, lambda: tf.image.flip_left_right(label), lambda: label
+    )
+    label = tf.cond(flip_ud > 0.5, lambda: tf.image.flip_up_down(label), lambda: label)
+
+    label = tf.squeeze(label, axis=-1)
+
+    contrast_factor = tf.random.uniform([], 0.5, 1.5)
+    brightness_factor = tf.random.uniform([], -0.2, 0.2)
+    img = tf.clip_by_value(img * contrast_factor + brightness_factor, 0.0, 1.0)
+
+    return img, label
+
+
+class TrainTestValidationDataset(ABC):
+    """A dataset that can be split into train, test, and validation sets."""
+
+    def __init__(self, train_split=0.8, val_split=0.1, test_split=0.1):
+        if train_split + val_split + test_split != 1.0:
+            raise ValueError("train_split + val_split + test_split must equal 1.0")
+
+        self.train_split = train_split
+        self.val_split = val_split
+        self.test_split = test_split
+
+        self.n_train = int(len(self) * self.train_split)
+        self.n_val = int(len(self) * self.val_split)
+        self.n_test = len(self) - self.n_train - self.n_val
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total number of samples in the dataset."""
+        ...
+
+    @abstractmethod
+    def _load(self, **kwargs):
+        """Load the full dataset."""
+        ...
+
+    def load_train(self, **kwargs):
+        return self._load(**kwargs).take(self.n_train)
+
+    def load_val(self, **kwargs):
+        return self._load(**kwargs).skip(self.n_train).take(self.n_val)
+
+    def load_test(self, **kwargs):
+        return self._load(**kwargs).skip(self.n_train + self.n_val).take(self.n_test)

```

Comment:
No, this is a great point! Shuffling only happens after the split, prior to training. I just did a [quick test](https://code.earthengine.google.com/e17e50be2bad862e0fac9d56777081d5) and the points produced by `ee.Image.sample` are sorted spatially, so by doing the splits sequentially we're segregating our train, test, and validation sets. I was actually able to drop our validation MSE for canopy cover in 2016 from ~139 to ~122 by shuffling the dataset before splitting it (no improvement in R2). Whether that's actually improving accuracy by allowing our model to see more representative training data or inflating performance by validating with adjacent, autocorrelated footprints might be debatable...

Any thoughts on what the best approach is here? Shuffling is an easy solution, but we might also consider training with some datasets and validating with others, if generalization across years is a high priority. Curious to hear your thoughts! We could also pose this question to the larger group.

Reply:
Actually, the question of autocorrelation crossed my mind.  The variable `MIN_SPACING` in `01_sampling.ipynb` is used in the call to `ee.Image.sample` which means that the *points* themselves are 150m apart, correct?  Does that in turn mean that the extracted footprints completely tesselate (but do not overlap) the image itself?  And how would this play in to the idea you have in #12 when image chips *would* overlap?  I obviously know next to nothing about CNNs, but is it fairly typical that the "sample" is pretty much the entire population?

> Any thoughts on what the best approach is here? Shuffling is an easy solution, but we might also consider training with some datasets and validating with others, if generalization across years is a high priority. Curious to hear your thoughts! We could also pose this question to the larger group.

Two thoughts:
1. I think that shuffling seems appropriate for the reasons you mention
2. I definitely think this is a great question for the larger group.  I know they would like to see what you've done here.

Reply:
> The variable MIN_SPACING in 01_sampling.ipynb is used in the call to ee.Image.sample which means that the points themselves are 150m apart, correct? Does that in turn mean that the extracted footprints completely tesselate (but do not overlap) the image itself? And how would this play in to the idea you have in https://github.com/lemma-osu/naip-cnn/issues/12 when image chips would overlap?

That's right, our datasets theoretically cover 100% of the LiDAR acquisitions, barring masked samples. In regards to #12, that overlap would only be used for generating predictions to reduce edge effects between chips, so it wouldn't affect model training or evaluation.

> is it fairly typical that the "sample" is pretty much the entire population?

My understanding is that it is pretty typical to use all available data, at least for LULC models, given that labeled data is usually hard to come by and CNNs are data-hungry. Since we *aren't* as data-limited, I imagine we could probably increase our spacing without any issues.

> I definitely think this is a great question for the larger group. I know they would like to see what you've done here.

Great! I figured this PR is a big enough step forward that it would be worth meeting up again to share progress and get everyone's thoughts, so I'll send out an email shortly and keep this question in mind.

For now though, it sounds like we're happy moving forward with shuffling?

Reply:
> For now though, it sounds like we're happy moving forward with shuffling?

Sounds great to me.